src/infra/src/table/migration/m20241114_000001_create_folder_table.rs

```diff
@@ -0,0 +1,216 @@
+use sea_orm_migration::prelude::*;
+
+#[derive(DeriveMigrationName)]
+pub struct Migration;
+
+const FOLDER_ORG_IDX: &str = "folder_org_idx";
+const FOLDER_FOLDER_ID_IDX: &str = "folder_folder_id_idx";
+const FOLDER_ORG_NAME_TYPE_IDX: &str = "folder_org_name_type_idx";
+
+#[async_trait::async_trait]
+impl MigrationTrait for Migration {
+    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .create_table(create_folder_table_statement())
+            .await?;
+        manager.create_index(create_folder_org_idx_stmnt()).await?;
+        manager
+            .create_index(create_folder_folder_id_idx_stmnt())
+            .await?;
+        manager
+            .create_index(create_folder_org_name_type_idx_stmnt())
+            .await?;
+        Ok(())
+    }
+
+    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .drop_index(Index::drop().name(FOLDER_ORG_NAME_TYPE_IDX).to_owned())
+            .await?;
+        manager
+            .drop_index(Index::drop().name(FOLDER_FOLDER_ID_IDX).to_owned())
+            .await?;
+        manager
+            .drop_index(Index::drop().name(FOLDER_ORG_IDX).to_owned())
+            .await?;
+        manager
+            .drop_table(Table::drop().table(Folder::Table).to_owned())
+            .await?;
+        Ok(())
+    }
+}
+
+/// Statement to create the folder table.
+fn create_folder_table_statement() -> TableCreateStatement {
+    Table::create()
+        .table(Folder::Table)
+        .if_not_exists()
+        .col(
+            ColumnDef::new(Folder::Id)
+                .integer()
+                .not_null()
+                .auto_increment()
+                .primary_key(),
+        )
+        // Surrogate key for folders that at the time of writing is...
+        // - stored in meta.key2
+        // - set by the server in calls to CreateFolder
+        // - referenced by the client in calls to ListDashboards
+        .col(ColumnDef::new(Folder::FolderId).string_len(256).not_null())
```

Comment:
We don't actually need both `id` and `folder_id`. `folder_id` would function as the primary key of the table.

Since users might have saved URLs containing the `folder_id` as a query param, and since dashboards inside the `meta` table still reference the `folder_id`, I think we should keep the `folder_id` around.

We could potentially get rid of the `id` column entirely. However, this would mean that when we eventually create a `dashboard` table that references `folder`, the foreign key on the `dashboard` table would need to reference the `folder.folder_id` column. Admittedly, I don't know if there are any negative performance ramifications to using the `varchar` `folder_id` column vs the `integer` `id` column. If joining on a `varchar` column has performance problems then this could be a reason for preferring to keep the `id` column and user it for foreign keys and joins.

Reply:
Another open question I have is how we want to store the `folder_id`. This ID is a "snowflake" generated by [ider.rs](https://github.com/openobserve/openobserve/blob/98429ca0fa59ae0b8b00ee842fdcb608e732c66f/src/config/src/ider.rs). I believe that the "snowflake" is actually a 64-bit integer so we could store it either as a varchar(256), a 64-bit int, or even a char(20).

A 64-bit int seems preferable to me if we can guarantee that it will, in fact, always be a 64-bit int.